### PR TITLE
Fix zero copy locks with hardlinks

### PR DESCRIPTION
**Pull request by tavplubix**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed a bug in zero-copy-replication (experimental feature) that could lead to data loss in rare cases when removing outdated mutated parts that previously failed to be committed to ZooKeeper

See https://github.com/ClickHouse/ClickHouse/pull/54641#issuecomment-1721445848

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit f8062d77e1a0cb271dc00cee7f9cdd514ce31049 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54859/f8062d77e1a0cb271dc00cee7f9cdd514ce31049/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

**Comment by tavplubix** [comment]
Stateless tests (release, DatabaseReplicated) [2/4] - https://github.com/ClickHouse/ClickHouse/issues/54894

File: src/Common/ZooKeeper/IKeeper.h [hunk 1]
```diff
@@ -510,6 +510,18 @@ class Exception : public DB::Exception
     const Error code;
 };
 
+class SimpleFaultInjection
+{
+public:
+    SimpleFaultInjection(Float64 probability_before, Float64 probability_after_, const String & description_);
+    ~SimpleFaultInjection() noexcept(false);
```
Comment (alesapin) [inline src/Common/ZooKeeper/IKeeper.h:517]: My c++ knowledge is bad... What does it mean `noexpect` with argument in method declaration?

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 2]
```diff
@@ -9247,8 +9252,19 @@ std::pair<bool, NameSet> StorageReplicatedMergeTree::unlockSharedDataByID(
...
-                zookeeper_part_node, part_name, children.size(), fmt::join(children, ", "));
+            /// It's possible that we have two instances of the same part with different blob names of
+            /// FILE_FOR_REFERENCES_CHECK aka checksums.txt aka part_unique_id,
+            /// and other files in both parts are hardlinks (the same blobs are shared between part instances).
+            /// It's possible after unsuccessful attempts to commit a mutated part to zk.
+            /// It's not a problem if we have found the mutation parent (so we have files_not_to_remove).
+            /// But in rare cases mutations parents could have been already removed (so we don't have the list of hardlinks).
+
+            /// I'm not 100% sure that parent_not_to_remove list cannot be incomplete (when it's not empty)
+            if (part_info.mutation && parent_not_to_remove.empty())
```
Comment (alesapin) [inline src/Storages/StorageReplicatedMergeTree.cpp:9263]: I see the comment, but still doesn't understand, sorry. 

I expected to see here some straightforward logic like check in zookeeper our "sibling" (this another part with different "FILE_FOR_REFERENCES_CHECK aka checksums.txt aka part_unique_id") hardlinks.

Minor: Why not just check `has_parent` flag? Why we explicitly check list of `parent_not_to_remove`?

File: src/Common/ZooKeeper/IKeeper.h (hunk 1 above)
Comment (tavplubix) [reply src/Common/ZooKeeper/IKeeper.h:517]: Dtors are noexcept by default, `noexcept(false)` disables that

File: src/Storages/StorageReplicatedMergeTree.cpp (hunk 2 above)
Comment (tavplubix) [reply src/Storages/StorageReplicatedMergeTree.cpp:9263]: The problem is that we don't have the list of hardlinks in the sibling part. We had it only in the parent part. And we have a problem if the parent part is already removed (we don't know the list of hardlinks in this case, so we have too keep all blobs)

> Minor: Why not just check has_parent flag?

Yes, looks like it's better to replace `parent_not_to_remove.empty()` with `!has_parent`

**Review by alesapin — APPROVED** [review]

**Pull request merged**